src/library_dylink.js

```diff
@@ -962,6 +965,12 @@ var LibraryDylink = {
 #if DYLINK_DEBUG
       err('dlsym: ' + symbol + ' getting table slot for: ' + result);
 #endif
+
+#if ASYNCIFY
```

Comment:
Hmm, sorry, I still don't understand this part. This is general linking code, so I'm surprised to see changes here - unless asyncify is using something we've never needed before, and I'm not sure what that is.

Also, the linked comment talks about functions, and below we call `addFunction`, but this checks `!= 0` on `.value`, which confuses me? Is it checking if it is not a function or something like that? If so then `typeof GOT[symbol]` might be better?

Reply:
I added this. Without it, addFunction will try to add the function which will at `setWasmTableEntry` because it's a JS function  and then triggers the assert a few lines later because it doesn't have a signature:

```
Aborted(Assertion failed: Missing signature argument to addFunction: function() {
                Asyncify.exportCallStack.push(x);
                try {
                  return original.apply(null, arguments);
                } finally {
                  if (!ABORT) {
                    var y = Asyncify.exportCallStack.pop();
                    assert(y === x);
                    Asyncify.maybeStopUnwind();
                  }
                }
```

When I retested this just now, I did however notice that just checking whether the symbol is in `GOT` (as proposed by me earlier) doesn't work if we have multiple libraries with the same symbols (similar issue to #16581 )

Reply:
Ah, then maybe we need to add the signature for the function? As long as the function has one, the existing asyncify code should copy it even into the wrapper function (the one that does `original.apply` that you pasted here),

https://github.com/emscripten-core/emscripten/blob/651e76db65712d32c61cbcd8cd1488e22ad2a5f5/src/library_async.js#L97-L102

Adding the signature would look like this:
```diff
diff --git a/src/library.js b/src/library.js
index b1947fc69..8e4c1f66e 100644
--- a/src/library.js
+++ b/src/library.js
@@ -3294,6 +3294,7 @@ LibraryManager.library = {
 
   $setWasmTableEntry__internal: true,
   $setWasmTableEntry__deps: ['$wasmTableMirror'],
+  $setWasmTableEntry__sig: 'ir',
   $setWasmTableEntry: function(idx, func) {
     wasmTable.set(idx, func);
     wasmTableMirror[idx] = func;
```
However the second param is a reference. I wrote "r" for that, in the diff. @sbc100 do we already have a letter for reference types?

Taking a step back, though, why does that function end up in a call to `addFunction()`? C/C++ code wouldn't be calling it, so I'm confused.

Reply:
`$setWasmTableEntry` is a JS-only function, it cannot be called from native code or added to the wasm table.. so I don't think it can/should have a signature at all.

Reply:
Sorry, I think a typo swallowed parts of my previous comments. The function being added isn't `setWasmTableEntry` - `setWasmTableEntry` is where the adding fails because no signature can be found. I clarified the above comment with bold/italic changes.

I attached the debugger at this point in question. You'll see that the function doesn't have a signature and is already added to GOT. (edit: I quickly checked -  it's been instrumented by `instrumentWasmExports` - this is line 17172 you see as `FunctionLocation`)
![image](https://user-images.githubusercontent.com/748698/164894784-ae77ddf2-3326-46de-8c8b-99392be4cf2f.png)

Reply:
> This approach though fails if multiple libraries have the same symbols.

1. Do you mean multiple libraries implement the same symbols? That would be incorrect, I think, as we need to declare this asyncify code just once (in the main module, and import it from side modules).
2. Or do you mean multiple libraries define the symbol (and all import it from one place)? That's the common case in general, which should be heavily tested.

Because of these questions I'm not sure why this PR is hitting a new type of problem in dynamic linking, so I'm still missing something...

> I was experimenting with keeping a reference to the original/wasm as part of the instrumented method. See https://github.com/chkuendig/emscripten/commit/2ea11b4734de5656aaf3285145831e2bb9ad53ae. But that also seems a bit hackish.

Interesting. I think that might make sense, actually - doesn't seem like a bad hack to me. But I still don't see why it is needed. So, we do copy signatures to the asyncify wrappers, here:

https://github.com/emscripten-core/emscripten/blob/3a3f3579d09c01ceb9616c90339b9064206dc747/src/library_async.js#L97-L102

It seems like you need the opposite, for exports as opposed to imports... do you know why that is, what is new in this PR that needs that?

Reply:
The changes, <https://github.com/chkuendig/emscripten/commit/2ea11b4734de5656aaf3285145831e2bb9ad53ae>, look good for me. It seems that `orig` can be passed to `addFunction` and `dlsym_js` expects that function index are returned.

```diff
#if ASYNCIFY
      if ('orig' in result) {
-       result = result.orig;
+       result = addFunction(result.orig);
      }
#endif
```

Reply:
https://github.com/emscripten-core/emscripten/blob/3a3f3579d09c01ceb9616c90339b9064206dc747/src/library_async.js#L97-L102

This is only enabled with the flags `ASSERTION`, `ASYNCIFY` and `MAIN_MODULE`, and can be applied to JavaScript function imports of main module. I think that this is no effect to imported functions from side modules, because they don't have .sig property.

Reply:
Ok, thanks, I re-read this now and I think I understand. Asyncify adds an extra JS wrapper around exports, and dynamic linking needs to see through that wrapper, basically. sgtm